python/cudf/cudf/core/frame.py

```diff
@@ -3505,6 +3510,226 @@ def to_arrow(self):
         """
         return self._column.to_arrow()
 
+    @property
+    def _copy_construct_defaults(self):
+        """A default dictionary of kwargs to be used for copy construction."""
+        raise NotImplementedError
+
+    def _copy_construct(self, **kwargs):
+        """Shallow copy this object by replacing certain ctor args.
+        """
+        return self.__class__(**{**self._copy_construct_defaults, **kwargs})
+
+    def _binaryop(
+        self,
+        other,
+        fn,
+        fill_value=None,
+        reflect=False,
+        lhs=None,
+        *args,
+        **kwargs,
+    ):
+        """Perform a binary operation between two single column frames.
+
+        Parameters
+        ----------
+        other : SingleColumnFrame
+            The second operand.
+        fn : str
+            The operation
+        fill_value : Any, default None
+            The value to replace null values with. If ``None``, nulls are not
+            filled before the operation.
+        reflect : bool, default False
+            If ``True`` the operation is reflected (i.e whether to swap the
+            left and right operands).
+        lhs : SingleColumnFrame, default None
+            The left hand operand. If ``None``, self is used. This parameter
+            allows child classes to preprocess the inputs if necessary.
+
+        Returns
+        -------
+        SingleColumnFrame
+            A new instance containing the result of the operation.
+        """
+        if lhs is None:
+            lhs = self
+
+        rhs = self._normalize_binop_value(other)
+
+        truediv_int_dtype_corrections = {
+            "int8": "float32",
+            "int16": "float32",
+            "int32": "float32",
+            "int64": "float64",
+            "uint8": "float32",
+            "uint16": "float32",
+            "uint32": "float64",
+            "uint64": "float64",
+            "bool": "float32",
+            "int": "float",
+        }
+
+        if fn == "truediv":
+            truediv_type = truediv_int_dtype_corrections.get(str(lhs.dtype))
+            if truediv_type is not None:
+                lhs = lhs.astype(truediv_type)
+
+        output_mask = None
+        if fill_value is not None:
+            if is_scalar(rhs):
+                if lhs.nullable:
+                    lhs = lhs.fillna(fill_value)
+            else:
+                # If both columns are nullable, pandas semantics dictate that
+                # nulls that are present in both lhs and rhs are not filled.
+                if lhs.nullable and rhs.nullable:
+                    # Note: using bitwise and rather than logical, but these
+                    # are boolean outputs so should be fine.
+                    output_mask = lhs.isnull() & rhs.isnull()
```

Comment:
This is really expensive because you're expanding the bitmask to a bytemask for the `isnull` operation. I think you want a `bitwise AND` operation instead against to two validity masks?

Reply:
I agree, that would be preferred but I wasn't sure how to go about doing that. The old code didn't call `isnull`, but it effectively did the same thing manually by constructing two `Series` objects, performing a `bitwise AND`, then extracting the underlying data buffer. Therefore I don't think this is a performance _regression_. Is there any Python/Cython API to perform bitwise operations directly on the data buffer? If so then I can definitely do that here, but if not I think changing this bit is probably out of scope for this PR since it would require exposing (and perhaps even implementing) a new API for buffers.

Reply:
IIRC the previous code constructed a `Series` around the validity buffers and then ran the `bitwise AND`, whereas you're doing extra GPU work here of expanding the validity buffers to boolean columns, then running a logical AND, then scattering null values where the boolean column is `True`.

I would expect we should be able to construct a `Column` using the validity buffer as the data buffer for the column and use those columns to run the `bitwise AND`. Then extract the data buffer from the result column and set that as the validity buffer on the output column.

Reply:
Just to close the loop here, I previously didn't realize that constructing a `Series` from a `cudf.core.buffer.Buffer` object (the nullmask) would result in a zero-copy construction on which `bitwise AND` would give the desired output. I'll reinstate the original code here to recover the performance. Note that in benchmarks I saw what appeared to be a 2-3% performance drop, but it was small enough that I chalked it up to noise. I'll double-check that it goes away here.